python/lsst/obs/pfs/pfsMapper.py

```diff
@@ -173,6 +176,128 @@ def _shiftAmpPixels(self, rawExp):
+            additionalData = butlerLocation.getAdditionalData()
+            locationString = butlerLocation.getLocations()[0]
+            locStringWithRoot = os.path.join(butlerLocation.getStorage().root, locationString)
+            logLoc = LogicalLocation(locStringWithRoot, additionalData)
+            # test for existence of file, ignoring trailing [...]
+            # because that can specify the HDU or other information
+            filePath = re.sub(r"(\.fits(.[a-zA-Z0-9]+)?)(\[.+\])$", r"\1", logLoc.locString())
+
+            if not os.path.exists(filePath):  # handle reruns
+                cfg = butlerLocation.storage.getRepositoryCfg(butlerLocation.storage.root)
+                for p in cfg.parents:
+                    nfilePath = os.path.join(p.root, locationString)
+                    if os.path.exists(nfilePath):
+                        filePath = nfilePath
+                        break
+
+            if checkExistence:             # we only want to know if the file exists
+                return os.path.exists(filePath)
+
+            if not os.path.exists(filePath):
+                raise RuntimeError("No such FITS file: " + logLoc.locString())
+
+            if getMetadata:
+                from lsst.afw.fits import readMetadata
+                return readMetadata(filePath, hdu=0)
+
+            if additionalData.get("llcX"):  # we were asked for a subimage
+                bbox = BoxI(PointI(additionalData["llcX"], additionalData["llcY"]),
+                            ExtentI(additionalData["width"], additionalData["height"]))
+                return afwImage.DecoratedImageF(filePath, hdu=hdu, bbox=bbox)
+            else:
+                return afwImage.DecoratedImageF(filePath, hdu=hdu)
+
+        if checkExistence:
+            return readData(datasetType, butlerLocation, dataId, checkExistence=True)
+
+        md = readData(datasetType, butlerLocation, dataId, getMetadata=True)
+        nread = md.get("W_H4NRED")
+        hasResetFrame = md["W_4FMTVR"] >= 3   # extra IMG/REF HDUs for the initial reset frame
+
+        hdu_img0 = 2 if hasResetFrame else 0   # hdu for first real read
+        #
+        # Process the data using CDS
+        #
+        data = {}
+        hdus = 2*np.array([1] if nread == 1 else [1, nread]) - 1
+        if "hdu" in dataId:
+            hdu = dataId["hdu"]
+            read = (hdu + 1)//2         # which read up the ramp do we want?
+            if read > nread:
+                raise RuntimeError(f"Requested read {read} (HDU {hdu}) is greater than W_H4NRED == {nread}")
+
+            hdus[-1] = hdu
+
+        if hdus[-1]%2 == 0:
+            raise RuntimeError(f"HDU {hdus[-1]} is a reference HDU")
+
+        for hdu in hdus:
+            read = (hdu + 1)//2         # which read up the ramp do we want?
+            try:
+                data[hdu] = readData(datasetType, butlerLocation, dataId, hdu=hdu + hdu_img0)
```

Comment:
Multiple calls to `readData` seem like it's doing the same work over and over. Surely it would be better to get a filename back from `readData` and then operate on that? I think `readData` should be split into a few pieces. It's not the extra work so much as the needless repeated filesystem checks that scare me. But this is all going away soon...

Reply:
I think we only read the pixels once.  The file system checks should be very subdominant; I think we should wait for Gen3